added pwrites

GTimothee opened the pull request:
-added pwrite in the writing parts of imageutils in order not to use seek+write
-changed name of reconstruct_img to "merge" to be consistent with the publication
-refactor some parts of the tests
-change the tests of merging parts to compare the content of the output nibabel images instead of comparing the hashes of the output files
-added python3.7 to yaml file

Commit:
adding pwrites

Commit:
add python3.7

Commit:
test travis

Commit:
test python37

Commit:
test write array

Commit:
test pwrite

Commit:
test pwrite

Commit:
/

Commit:
/

Commit:
test header

Commit:
test header

Commit:
test header

Commit:
test header

Commit:
test header

Commit:
test header

Commit:
test header

Commit:
fix

Commit:
fix

Commit:
test

Commit:
test

Commit:
test

Commit:
test

Commit:
remove commented code

Commit:
add header assertion to tests

Commit:
remove virtualenv

glatard commented:
Hi @GTimothee, thanks for this! The code looks good to me but I'll leave it to @ValHayot to approve it. Could you perhaps add a line of explanation in the PR description?

GTimothee commented:
Yes sure, it is done.

ValHayot commented on .travis.yml:
```diff
@@ -7,13 +7,17 @@ python:
 script:
-    - pycodestyle sam/imageutils.py tests/*.py
-    - pytest --cov=sam ./
+    - python -m pytest
```
I know there is better that the way i used, but coveralls no longer works. Also, is there a reason for removing pycodestyle?

GTimothee replied:
The reason is that pycodestyle failed I don't know why so I removed it (it was just telling me that it failed with code 1 without other explanations). I also removed --cov because I did not know what it was and I just wanted to make it work. But we can add --cov back I just did not remember to do it.

ValHayot commented on .travis.yml:
```diff
@@ -7,13 +7,17 @@ python:
+matrix:
+  include:
+    - python: 3.7
+      dist: xenial
```
Not that it's really important right now, but is there a reason for using a different distro of ubuntu for the python version? I think i'd prefer that either all tests are done on the same distro or that we redo each test for each distro. You can set the environment of all the tests to Xenial if you prefer, however

GTimothee replied:
It is just because the default did not found a package for python3.7. But as it is working like other python versions maybe we can remove it. I just wanted to use python3.7 because it is the one I used while developing.

ValHayot commented on sam/imageutils.py:
```diff
@@ -1514,15 +1515,12 @@ def write_array_to_file(data_array, to_file, write_offset, hdfs_client=None):
-            os.fsync(f)
-            write_time += time() - write_start
+
+        fd=os.open(to_file, os.O_RDWR | os.O_APPEND)
```
You removed the seek increment here. Are we considering seeks to only be the system call that increments the counter? Otherwise, seeks are still occurring. This, of course, is not important for the practical use of `sam`, but for benchmarking it would be.

Also, a perhaps misplaced comment: for your benchmarking did you have `os.fsync()` in clustered, multiple and naive but not in pwrites? that might explain at least a portion of the speedup.

GTimothee replied:
I did not use fsync in my benchmarking. For the count of seeks yes I thought about nb_seeks as the number of actual system calls to seek. But anyway I wanted to recheck all those counts and profiling in the week for my future tests.

GTimothee replied:
Also I plan to change the output value of the split and merge functions because the different functions return different values etc. I thought of a python dict but if you don't want it can still be a tuple.

ValHayot commented on sam/imageutils.py:
```diff
@@ -1546,24 +1544,14 @@ def write_dict_to_file(data_dict, to_file, bytes_per_voxel, header_offset):
-            # print "seek point:", seek_pos, to_file.tell()
-            seek_start = time()
-            to_file.seek(seek_pos, 0)
-            seek_number += 1
```
i think this might still be relevant here, for benchmarking anyway, depending on if we define a seek to be a disk seek or just the os system call. @glatard thoughts?

ValHayot reviewed (changes requested):
Hey @GTimothee, looks really good. Just some minor comments, most of which you can probably ignore.

Commit:
return dict

Commit:
remove hdfs support, add comments and add robustness in init function of imageutils class

Commit:
remove hdfs support, add comments and add robustness in init function of imageutils class

Commit:
remove hdfs support, add comments and add robustness in init function of imageutils class

GTimothee committed:
added seek counts to naive split, restored seek counts in writing parts

GTimothee committed:
refactoring of clustered writes

GTimothee committed:
refactor split_multiple_writes

GTimothee committed:
review benchmarking clustered read

GTimothee committed:
remove unuseful code

GTimothee committed:
remove unuseful code

GTimothee committed:
inspect multiple_read banchmarking

GTimothee committed:
inspect multiple_read banchmarking

GTimothee committed:
adding changes to travis

GTimothee commented:
Hi, I updated the code so I think you can merge it now. Please tell me what to change if not. The code passed the tests but I just added back the pycodestyle code and it fails with the following error code "The command "pycodestyle sam/imageutils.py tests/*.py" exited with 1." I don't know why. As the only difference between the working commit and the failing commit is the pycodestyle line, it should pass the test on the TravisCI of the project as it has been successfully used so far.

ValHayot commented:
hey @GTimothee, if you look right above the line that says, "The command "pycodestyle sam/imageutils.py tests/*.py" exited with 1.", it will detail why pycodestyle failed. Seems to be some whitespace issues amongst other things. Also looks like the test cases themselves are no longer working.

GTimothee committed:
fix float issues for centos

GTimothee committed:
pycodestyle

GTimothee committed:
pycodestyle

GTimothee committed:
fix importerror

GTimothee commented:
Updated

ValHayot reviewed (approved):
Awesome! This looks really great @GTimothee.

GTimothee commented:
Thank you :) (sorry I misclicked so I closed the pull request accidentally)
